[CA-1628] Create Storage Transfer Job code

s-rubenstein opened the pull request:
RR: https://broadworkbench.atlassian.net/browse/CA-1628

In this change, @s-rubenstein and I have added functionality to create storage transfer servive jobs to transfer the contents of one workspace bucket to another.
To do this, we've added a new table called "PPW_STORAGE_TRANSFER_SERVICE_JOB" that tracks STS jobs related to workspace migration attempts. The idea is we create a job and write into the database and then poll sts for updates on each job (we could write a bespoke actor for this). When the job is done, we move to the next stage of migration.

This change seems a little large and most of the noise has been generated from moving things around. In this change:
- I decided that the "WorkspaceMigrationMonitor" was getting too large and moved the database-related types and functions into their own files, private to the newly created "migration" package. 
- I dropped the "V1" prefix because it's unlikely that we'll still be migrating v1 workspaces by the time v3 comes along
- I eliminated the use of behavioural inheritance in these types and hid the specifics of table "truncation" in the monitor. Inheritance is evil. Inheritance must die.

RR: https://broadworkbench.atlassian.net/browse/CA-1629
Add new timestamp to database to record when this occurs.
Add some new extension methods to better support chaining via the `.`
operator.

RR: https://broadworkbench.atlassian.net/browse/CA-1632
Add timestamp to migration table to record when the bucket was deleted.
Some minor fixes, including ensuring that buckets are deleted
recursively.

RR: https://broadworkbench.atlassian.net/browse/CA-1631
Re-use most of Sky's code for transferring to the temp bucket.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

s-rubenstein committed:
Fix fk length

ehigham committed:
Create Storage Transfer Job code

ehigham committed:
extract db operations into their own files + eliminate inheritance

ehigham committed:
add unit test

ehigham committed:
tidy up

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/Shared.scala:
```diff
@@ -0,0 +1,33 @@
+import cats.implicits.{catsSyntaxOptionId, toTraverseOps}
+import org.broadinstitute.dsde.rawls.RawlsException
+
+object Shared {
```
rename to something more informative/specific. Maybe `MigrationHelpers`?

ehigham replied:
Sure

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationMonitor.scala:
```diff
@@ -298,22 +174,44 @@ object V1WorkspaceMigrationMonitor
+                                          destBucket: GcsBucketName,
+                                          projectToBill: GoogleProject,
+                                          sts: GoogleStorageTransferService[IO]
+                                          ): IO[(TransferJob, WriteAction[Unit])] =
```
I'm starting to think that this return type is not what we want. Returning the `DBIOAction` separately means we aren't guaranteed to write the result of the STS operation (or any other GCP operations) to the DB immediately after they complete -- it leaves it up to whoever is calling this fn to be diligent about running the `DBIOAction` right after and leaves open the possibility that they won't run it at all which we definitely don't want. The alternative is to wrap all calls to the `MigrationMonitor` in `inTransaction` blocks, but that's not ideal either since a lot of those calls involve issuing requests to GCP which would tie up DB connections. 

My proposed solution would be to run the `DBIOAction` in this function and have the return type just be `IO[TransferJob]`.

ehigham replied:
I understand the concern but I worry that'll lead to mixing/nesting IO and database transactions and will cause all sorts of pain. At least this way it's harder to do that. While it maybe be up to the caller to get it right, we'll be the ones writing that code so it'll be fairly straightforward to prevent (ignoring return values usually indicates smell). I'll think about it, but I'm not going to do this just yet.

ehigham committed:
rename Shared -> MigrationUtils

ehigham committed:
[CA-1629][CA-1631][CA-1632] Delete Workspace Buckets, Transfer Contents of Tmp Bucket to Final Workspace Bucket (#1593)

RR: https://broadworkbench.atlassian.net/browse/CA-1629
Add new timestamp to database to record when this occurs.
Add some new extension methods to better support chaining via the `.`
operator.

RR: https://broadworkbench.atlassian.net/browse/CA-1632
Add timestamp to migration table to record when the bucket was deleted.
Some minor fixes, including ensuring that buckets are deleted
recursively.

RR: https://broadworkbench.atlassian.net/browse/CA-1631
Re-use most of Sky's code for transferring to the temp bucket.

s-rubenstein reviewed (commented):
Thumbs up for all Edmund's changes on top of mine! I am PR author so can't approve myself.

ehigham reviewed (approved):
Approval on behalf of @s-rubenstein

ehigham committed:
`msuccess` to `successOpt`

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/monitor/WorkspaceMigrationMonitorSpec.scala:
```diff
@@ -0,0 +1,413 @@
+        .head
+        .copy(
+          newGoogleProjectId = GoogleProjectId(destProject).some,
+          tmpBucketName = GcsBucketName(sourceBucket).some
```
This might seem trivial, but can we just use another bucket as the `tmpBucket` here?  I know it technically works for the test but in reviewing it, I had to stop and try to figure out if there was a good reason why it was using `sourceBucket` and it doesn't seem like there is.

```suggestion
          tmpBucketName = GcsBucketName("foo-bar").some
```

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/monitor/WorkspaceMigrationMonitorSpec.scala:
```diff
@@ -0,0 +1,413 @@
+  "deleteTemporaryBucket" should "delete the temporary bucket and record when it was deleted" in {
+    val sourceProject = "general-dev-billing-account"
+    val sourceBucket = "az-leotest"
+    val destProject = "terra-dev-7af423b8"
```
I think the above "real" values are artifacts from when were were actually testing this thing against real buckets.  Now that this test is a proper unit test again, can we change these values to ones that are clearly "test" values?  I know it technically works, but I don't want future maintainers to come in here and think this test needs very specific values or projects to be used here.

ehigham replied:
Oh perhaps it shouldnt be...

ehigham replied:
I've removed all the redefinitions - they weren't needed

ehigham committed:
remove unused test val redefinitions

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/monitor/WorkspaceMigrationMonitorSpec.scala:
```diff
@@ -0,0 +1,389 @@
+  "createFinalBucket" should "create a new bucket in the same region as the tmp workspace bucket" ignore {
+    val sourceProject = "general-dev-billing-account"
+    val sourceBucket = "az-leotest"
+    val destProject = "terra-dev-7af423b8"
```
Similar to the other change, can we remove these "real" values?  There's another occurrence around line 111 too.

ehigham replied:
we actually use these when we run the test manually so not yet
